src/backend/distributed/transaction/multi_shard_transaction.c

```diff
@@ -103,80 +86,120 @@ CreateShardConnectionHash(void)
+ * all placements of a shard (specified by shard identifier). After sending a
+ * BEGIN command on all connections, they are added to shardConnectionHash for
+ * use within this transaction. Exits early if connections already exist for
+ * the specified shard, and errors if no placements can be found, a connection
+ * cannot be made, or if the BEGIN command fails.
  */
 void
-OpenConnectionsToShardPlacements(uint64 shardId, HTAB *shardConnectionHash,
-								 char *userName)
+BeginTransactionOnShardPlacements(uint64 shardId, char *userName)
 {
-	bool shardConnectionsFound = false;
-
-	/* get existing connections to the shard placements, if any */
-	ShardConnections *shardConnections = GetShardConnections(shardConnectionHash,
-															 shardId,
-															 &shardConnectionsFound);
+	List *shardPlacementList = NIL;
+	ListCell *placementCell = NULL;
 
-	List *shardPlacementList = FinalizedShardPlacementList(shardId);
-	ListCell *shardPlacementCell = NULL;
-	List *connectionList = NIL;
+	ShardConnections *shardConnections = NULL;
+	bool shardConnectionsFound = false;
 
-	Assert(!shardConnectionsFound);
+	MemoryContext oldContext = NULL;
 
+	shardPlacementList = FinalizedShardPlacementList(shardId);
 	if (shardPlacementList == NIL)
 	{
+		/* going to have to have some placements to do any work */
 		ereport(ERROR, (errmsg("could not find any shard placements for the shard "
 							   UINT64_FORMAT, shardId)));
 	}
 
-	foreach(shardPlacementCell, shardPlacementList)
+	/* get existing connections to the shard placements, if any */
+	shardConnections = GetShardConnections(shardId, &shardConnectionsFound);
+	if (shardConnectionsFound)
 	{
-		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(
-			shardPlacementCell);
-		char *workerName = shardPlacement->nodeName;
-		uint32 workerPort = shardPlacement->nodePort;
-		PGconn *connection = ConnectToNode(workerName, workerPort, userName);
+		/* exit early if we've already established shard transactions */
+		return;
+	}
+
+	foreach(placementCell, shardPlacementList)
+	{
+		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(placementCell);
+		XactModificationType oldModLevel = XactModificationLevel;
+		PGconn *connection = NULL;
 		TransactionConnection *transactionConnection = NULL;
+		PGresult *result = NULL;
+
+		/*
+		 * This is extremely hacky, but allows adding connections for new shard
+		 * placements during a pure-DDL transaction block. Without this, we'd
+		 * get an error from ConnectToNode about not permitting new connections
+		 * after the first modification in a transaction block. We couldn't get
+		 * to this point if data modifications have already occurred (because
+		 * of the check in ExecuteDistributedDDLCommand), so we know at most
+		 * only schema modifications have already happened.
+		 *
+		 * FIXME: Remove this hack during big connection/transaction revision
+		 */
+		XactModificationLevel = XACT_MODIFICATION_INVALID;
```

Comment:
> maybe we can add a flag to ConnectToNode to override the check?

Expanding the API in that way _codifies_ the hackery I have in place here, no? In this case, it's really ugly to do the thing I'm doing; in that case, it's part of the API (i.e. encouraged).

I think removing the check from `ConnectToNode` is our best bet; it can probably be placed elsewhere without too many problems. But in the upcoming connection stuff, I think the connection library must centralize (and enforce) all this stuff, and no PR should be allowed to relax the restrictions (i.e. if you find something you can't do, bring it up as a new design enhancement to the connection/transaction API, no changing a method to permit something that hasn't been reasoned about in the grand scheme of "how we use connections").

I'll do the "remove the check" bit, and move the check into all the callers.